tools/fast/environment_devnet.go

```diff
@@ -0,0 +1,183 @@
+	}
+
+	if err := os.MkdirAll(env.location, 0775); err != nil {
+		return nil, err
+	}
+
+	return env, nil
+}
+
+// GenesisCar provides a url where the genesis file can be fetched from
+func (e *EnvironmentDevnet) GenesisCar() string {
+	uri := url.URL{
+		Host:   fmt.Sprintf("genesis.%s.kittyhawk.wtf", e.network),
+		Path:   "genesis.car",
+		Scheme: "https",
+	}
+
+	return uri.String()
+}
+
+// GenesisMiner returns a ErrNoGenesisMiner for this environment
+func (e *EnvironmentDevnet) GenesisMiner() (*GenesisMiner, error) {
+	return nil, ErrNoGenesisMiner
+}
+
+// Log returns the logger for the environment.
+func (e *EnvironmentDevnet) Log() logging.EventLogger {
+	return e.log
+}
+
+// NewProcess builds a iptb process of the given type and options passed. The
+// process is tracked by the environment and returned.
+func (e *EnvironmentDevnet) NewProcess(ctx context.Context, processType string, options map[string]string, eo EnvironmentOpts) (*Filecoin, error) {
+	e.processesMu.Lock()
+	defer e.processesMu.Unlock()
+
+	ns := iptb.NodeSpec{
+		Type:  processType,
+		Dir:   fmt.Sprintf("%s/%d", e.location, len(e.processes)),
+		Attrs: options,
+	}
+
+	e.log.Infof("New Process type: %s, dir: %s", processType, ns.Dir)
+
+	if err := os.MkdirAll(ns.Dir, 0775); err != nil {
+		return nil, err
+	}
+
+	c, err := ns.Load()
+	if err != nil {
+		return nil, err
+	}
+
+	// We require a slightly more extended core interface
+	fc, ok := c.(IPTBCoreExt)
+	if !ok {
+		return nil, fmt.Errorf("%s does not implement the extended IPTB.Core interface IPTBCoreExt", processType)
+	}
+
+	p := NewFilecoinProcess(ctx, fc, eo)
+	e.processes = append(e.processes, p)
+	return p, nil
+}
+
+// Processes returns all processes the environment knows about.
+func (e *EnvironmentDevnet) Processes() []*Filecoin {
+	e.processesMu.Lock()
+	defer e.processesMu.Unlock()
+	return e.processes[:]
+}
+
+// Teardown stops all of the nodes and cleans up the environment.
+func (e *EnvironmentDevnet) Teardown(ctx context.Context) error {
+	e.processesMu.Lock()
+	defer e.processesMu.Unlock()
+
+	e.log.Info("Teardown environment")
+	for _, p := range e.processes {
+		if err := p.core.Stop(ctx); err != nil {
+			return err
+		}
+	}
+
+	return os.RemoveAll(e.location)
+}
+
+// TeardownProcess stops the running process and removes it from the
+// environment.
+func (e *EnvironmentDevnet) TeardownProcess(ctx context.Context, p *Filecoin) error {
+	e.processesMu.Lock()
+	defer e.processesMu.Unlock()
+
+	e.log.Infof("Teardown process: %s", p.core.String())
+	if err := p.core.Stop(ctx); err != nil {
+		return err
+	}
+
+	for i, n := range e.processes {
+		if n == p {
+			e.processes = append(e.processes[:i], e.processes[i+1:]...)
+			break
+		}
+	}
+
+	// remove the provess from the process list
+	return os.RemoveAll(p.core.Dir())
+}
+
+// GetFunds retrieves a fixed amount of tokens from an environment
+func GetFunds(ctx context.Context, env Environment, p *Filecoin) error {
```

Comment:
Would be interested in thoughts on how to get this to work with the MemoryGenesis environment. My first thought was another function which would set some value on the environment. Not sure if it's really needed. Would be nice to provide reusability though.

Reply:
What do you think of having the MemoryGenesis run a faucet process? We could reuse almost all of this code if that was the case.

Another idea would be having the MemoryGenesis keep a reference to the genesis node process and adding separate logic to pull funds from that.

Reply:
Ya, I think we'd need to have a way to attach a node to the environment, which could just be a stand alone function that would store a pointer to a `process`. I'll do that in a follow up issue I think. Once this merges I'll write it up.